src/storage/src/render/upsert.rs

```diff
@@ -295,10 +294,30 @@ where
         key_indices_sorted.clone(),
         &upsert_envelope.key_indices,
     );
+
+    // It's very important to hash the right thing. We have nested `Option` and
+    // `Result` here. And, for example, `Some(key).hashed()` is not the same as
+    // `key.hashed()`.
+    //
+    // We make sure  to hash the same thing for both previous updates and new
+    // updates.
+    //
+    // Also: this problem was only showing up when trying to use upsert-style
+    // sources with multiple storaged workers.
     let result_stream = stream.binary_frontier(
         &previous_ok.inner,
-        Exchange::new(move |DecodeResult { key, .. }| key.hashed()),
-        Exchange::new(|((key, _v), _t, _r)| Ok::<_, Infallible>(key).hashed()),
+        Exchange::new(move |DecodeResult { key, .. }| {
+            // N.B. We make the expected type explicit here to make sure it
+            // cannot change by accident.
+            let key: &Option<Result<Row, DecodeError>> = key;
+            key.hashed()
+        }),
+        Exchange::new(|((key, _v), _t, _r)| {
+            // N.B.  We make the expected type explicit here to make sure it
+            // cannot change by accident.
+            let key: &Result<Row, DecodeError> = key;
+            Some(key).hashed()
```

Comment:
this is hashing a `Option<Result<Row, DecodeError>>` and a `Option<&Result<Row, DecodeError>`

I know they are the same because of https://doc.rust-lang.org/src/core/hash/mod.rs.html#944-949, but if you want to be even more clear you can `as_ref` the first one